dataset - histology

Borda opened the pull request:
making the dataset readable and prepare for adding new dataset...

kaspermarstal committed:
BUG: Let SimpleITK write labels with world information (instead of manually generating mhd header); STYLE: Move more functions to util.py

kaspermarstal committed:
BUG: Try to warp images with SimpleITK instead of SuperElastix

Borda committed:
refactoring datasets

kaspermarstal commented:
Thank you for this massive contribution! 

Have you tested locally with any the data sets that experiments runs as expected? I will also test locally.

kaspermarstal commented on ContinuousRegistration/Source/datasets.py:
```diff
@@ -521,6 +557,9 @@ def __init__(self, input_directory, output_directory, max_number_of_registration
         self.evaluate_point_set(superelastix, file_names, output_directory)
+
+
+class Histology(Dataset):
```
Do you have a "code name" for your data set? We might add other histology data sets in the future

Borda replied:
not yet, but we can think about it :)

kaspermarstal replied:
Okay great! Probably better now that later, so that the arguments (e.g. --histology-input-directory) would have to be changed in the future :)

Borda commented:
Well, that is something I am missing some description how to run the tests or what is the data setting (I can't find it in readme nor documentation) so I was hoping to see the result from CI...Can I access the detail results from Jenkins?

kaspermarstal commented:
@N-Dekker what is the link to the cdash page?

kaspermarstal commented:
Example python scripts are here: https://continuousregistration.grand-challenge.org/documentation/. Let me know if this does not suffice!

Borda commented:
you have typo in instructions (the relative path):
`mkdir build && cd build && cmake ../SuperBuild && make -j4`

kaspermarstal commented:
What is the typo?

Borda commented:
the relative path seems not correct, remove `SuperElastix/`

kaspermarstal commented:
You are right, thanks

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1252869)

kaspermarstal commented:
Thanks Niels! Will investigate those failures.

@Borda you can merge develop now for a working Travis build

Borda commented:
@kaspermarstal done :) but does the Travis run also the python tests?

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1252914)

Borda commented:
@kaspermarstal I have added the python testing to Travis, see #226

kaspermarstal commented:
What an awesome contribution!

kaspermarstal commented:
Maybe we should also add a call to ctest to also run the C++ tests

Borda commented:
ok, let's discuss it in the related thread #226 and keep this for the "adding dataset"

Borda commented:
@kaspermarstal  I do not think that is a good idea to recommend users to make a build directory outside the SuperElastix (imagine there will be several projects asking the same so you would have a collision among builds) I would recommend to make the build in the directory and ignore any build in `.gitignore`

Borda committed:
add Histology dataset

Borda commented:
@kaspermarstal I have added the dataset class, it is what you expected?

kaspermarstal commented:
Yes that looks really good! Can you also add the command line arguments to https://github.com/SuperElastix/SuperElastix/blob/develop/ContinuousRegistration/Source/make_registration_scripts.py#L16-L25? 

And have you tested that everything runs end-to-end locally? You can just use https://github.com/SuperElastix/SuperElastix/blob/develop/ContinuousRegistration/Submissions/TeamElastix/ElastixAffineTransform.json and change every mention of `Dimensionality: 3` to `Dimentionality: 2`. With the `--blueprint-file-name` python option you can ignore all the other parameter maps.

Borda committed:
add HistoBIA as param

Borda commented:
@kaspermarstal well, not really, I was not even able to run the POPI according to the [documenttaion](https://continuousregistration.grand-challenge.org/documentation/) which should be similar to ours by using registration points...
For POPI 
```
python make_registration_scripts.py \
	--superelastix /mnt/home.dokt/borovji3/Repos/SuperElastix/build/SuperElastix-build/bin/SuperElastix \
	--submissions-directory /mnt/home.dokt/borovji3/Repos/SuperElastix/ContinuousRegistration/Submissions \
	--output-directory /datagrid/Medical/microscopy/TEMPORARY/CRC \
	--popi-input-directory /datagrid/Medical/microscopy/TEMPORARY/CRC/POPI \
	--max-number-of-registrations-per-dataset 4
```
I got:
```
MetaImage: Read: Cannot open data file
Traceback (most recent call last):
  File "/local/borovji3/Applications/PyCharm.Pro/helpers/pydev/pydevd.py", line 1664, in <module>
    main()
  File "/local/borovji3/Applications/PyCharm.Pro/helpers/pydev/pydevd.py", line 1658, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/local/borovji3/Applications/PyCharm.Pro/helpers/pydev/pydevd.py", line 1068, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/local/borovji3/Applications/PyCharm.Pro/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home.dokt/borovji3/Dropbox/Workspace/SuperElastix/ContinuousRegistration/Source/make_registration_scripts.py", line 213, in <module>
    run(parameters)
  File "/home.dokt/borovji3/Dropbox/Workspace/SuperElastix/ContinuousRegistration/Source/make_registration_scripts.py", line 137, in run
    datasets = load_datasets(parameters)
  File "/home.dokt/borovji3/Dropbox/Workspace/SuperElastix/ContinuousRegistration/Source/make_registration_scripts.py", line 109, in load_datasets
    parameters.max_number_of_registrations_per_dataset)
  File "/mnt/datagrid/personal/borovec/Dropbox/Workspace/SuperElastix/ContinuousRegistration/Source/datasets.py", line 492, in __init__
    displacement_field_file_names[0])),
  File "/mnt/datagrid/personal/borovec/Dropbox/Workspace/SuperElastix/ContinuousRegistration/Source/util.py", line 99, in create_mask_by_size
    image = sitk.ReadImage(image_file_name)
  File "/home.dokt/borovji3/.local/lib/python3.5/site-packages/SimpleITK/SimpleITK.py", line 8614, in ReadImage
    return _SimpleITK.ReadImage(*args)
RuntimeError: Exception thrown in SimpleITK ReadImage: /tmp/SimpleITK-build/ITK/Modules/IO/Meta/src/itkMetaImageIO.cxx:491:
itk::ERROR: MetaImageIO(0x56112a377780): File cannot be read: /datagrid/Medical/microscopy/TEMPORARY/CRC/POPI/bh/mhd/00.mhd for reading.
Reason: Success
```

and for HBIA
```
python make_registration_scripts.py \
	--superelastix /mnt/home.dokt/borovji3/Repos/SuperElastix/build/SuperElastix-build/bin/SuperElastix \
	--submissions-directory /mnt/home.dokt/borovji3/Repos/SuperElastix/ContinuousRegistration/Submissions \
	--output-directory /datagrid/Medical/microscopy/TEMPORARY/CRC \
	--hbia-input-directory /datagrid/Medical/dataset-histology-stain-slices \
	--max-number-of-registrations-per-dataset 4
```
I got:
```
ERROR:root:Dataset "POPI" requested by "ElastixBSplineTransform" but no directory provided. Skipping dataset for this blueprint. See '--help' for usage.
ERROR:root:Dataset "DIRLAB" requested by "ElastixBSplineTransform" but no directory provided. Skipping dataset for this blueprint. See '--help' for usage.
ERROR:root:Dataset "EMPIRE" requested by "ElastixBSplineTransform" but no directory provided. Skipping dataset for this blueprint. See '--help' for usage.
...
ERROR:root:Dataset "MGH10" requested by "ITKv4_SyN_CC" but no directory provided. Skipping dataset for this blueprint. See '--help' for usage.
ERROR:root:Dataset "SPREAD" requested by "ITKv4_SyN_CC" but no directory provided. Skipping dataset for this blueprint. See '--help' for usage.
```

kaspermarstal commented:
You can also send me some data and I can have a go? Cannot see from the log why it fails to read the image.

kaspermarstal commented:
Actually, you also need to provide mask file names, even if masks do not make sense for the data set. You can use 

    mask_file_names = [create_mask_by_thresholding(label_file_name_pair, displacement_field_file_name_pair, output_directory, -1., 0., 0.)
                       for label_file_name_pair, displacement_field_file_name_pair in zip(label_file_names, displacement_field_file_names)]

to just generate masks that cover the entire image.

kaspermarstal commented:
Also, you need to add the name "HistoBIA" to the "Dataset" variable in the blueprint. The second round of errors is just because no input directories are provided to the other data sets that appear in this variable in the blueprint.

Borda committed:
add empty mask @HistoBIA dataset

Borda commented:
@kaspermarstal I believe that I added the HBIA name everywhere now, but, please check it
I used the similar schema as for POPI and added an empty mask...
The link to data is in class description and also in the mail, I sent to CRC recently :)

kaspermarstal commented:
Allright, thanks!

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1258896)

Borda commented:
@kaspermarstal may I consider it as done from my side? :)

kaspermarstal commented:
Yes, it looks good, although I do think that you should run experiments locally, and make sure that everything works end-to-end, and that the validation points sets can also be used by the system etc. 

If you run into any problems, or have any questions, let me know and I will update the information on the website. Participants that uses your data will have to go through the same process, so let us make it as easy as possible for them!

Borda commented:
@kaspermarstal OK, I would like to start with one or you prepared datasets, POPI. I have downloaded the data in the same structure you have presented on your site, see [here](http://cmp.felk.cvut.cz/~borovji3/share/POPI.zip) what do I have wrong with data that cannot be read, can you read it?

kaspermarstal commented:
No, there is indeed something wrong, let me get back to you with that.

kaspermarstal commented:
Try to use this command: `wget --recursive --accept 00.mhd,00.raw,50.mhd,50.raw,*.pts --no-parent https://www.creatis.insa-lyon.fr/~srit/POPI/MedPhys11`

Borda committed:
fix loading HBIA dataset

Borda commented:
@kaspermarstal ok, that helps and no I am running the registration which takes quite some time...
well, I was able to run the HBIA and fixed some minor issues :)

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1260860)

kaspermarstal commented:
Great!

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1261196)

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1264714)

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1267299)

Borda commented:
@kaspermarstal @FBerendsen just a short question, do you see realistic merging this dataset to be able use it also in CRC challenge? :)

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1268984)

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1270476)

kaspermarstal commented:
Yes, almost ready for merge! Testing locally now. The field "displacement_field_file_names" needs to be defined in the init() function for the dataset. This is where the output deformation field is saved. It is a path relative to the output_directory. You can see in the other data sets how can be made.

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1270485)

Borda committed:
load CSV points

cdashbot-niels-dekker commented:
[PR-223;Borda;Test experienced failing tests.](http://trunk.cdash.org/viewTest.php?onlyfailed&buildid=1271231)

Borda commented:
for loading landmarks, I use pandas library, but somehow it does not run well on Travis with python3.4 and when I try to install python3.5 it says `E: Unable to locate package python3.5`

kaspermarstal commented:
OK. Is it possible to convert the landsmarks to .vtk format?

kaspermarstal commented:
Even though it is sub-optimal solution to convert to vtk, we should merge this as soon as possible.

kaspermarstal commented:
You can also disable testing of the files that causes the problems on Travis, add a TODO, and we can add them later.

Borda commented:
@kaspermarstal well, I wrote own loading of the CSV landmarks... now it seems that everything is passing

kaspermarstal commented:
Great! Can you make them available for me to test? Then we can merge today!

Borda commented:
@kaspermarstal what do you mean by available for you, the CSV landmarks are part of the dataset and loading function is in the utils module, see [L148](https://github.com/Borda/SuperElastix/blob/a2ab37a1836414533d0c89eca862111225bb2fa3/ContinuousRegistration/Source/util.py#L148)

kaspermarstal commented:
OK, no problem. I assumed you had had to generate new ground truth files and that these were not available yet.

kaspermarstal commented:
There are some problems loading the data for evaluation. I am preparing a patch.

Borda commented:
@kaspermarstal what kind of problem?

kaspermarstal commented:
Portions of the code are not setup to handle different file formats, e.g. https://github.com/SuperElastix/SuperElastix/blob/develop/ContinuousRegistration/Source/metrics.py#L13-L14. I believe the underlying issue is that we pass around file paths instead of in-memory data. I am currently moving all data reading into the data set classes where it belongs, so as not to leak this kind of data set specific code outside these classes.

Borda commented:
Be aware of passing too much data inside as parameters. By my experience with multiprocessing in python, it may crash if the data are too big and some `pickle` fix is required.
see http://stackoverflow.com/questions/25156768

kaspermarstal commented:
Ok, thanks for the heads up!

kaspermarstal commented:
@Borda can you merge the hbia branch in this repo? Not sure if I can push directly to your pull request,

kaspermarstal commented:
You can also give me acces to your fork then I can do it

Borda commented:
@kaspermarstal I have check option "Allow edits from maintainers." so you should have access to the PR and also I added you to my fork... or should I merge it?

kaspermarstal commented:
If you could merge it that would be a big help thanks

Borda commented:
@kaspermarstal just to clarify, merge this PR to develop branch, right?

kaspermarstal commented:
Yep, to develop!

kaspermarstal commented:
Did I merge this before you got a chance to port the changes from the hbia branch? There is a lot of code I wrote for your dataset to be able to run that is now missing from develop

Borda commented:
@kaspermarstal I am not sure, I just read mail with merge status so I thought that it is done... but from the message "kaspermarstal merged commit dc43751 into SuperElastix:develop" it seems that only one commit was merged, right?

Borda commented:
Now I see the `class HBIA(Dataset)` in datasets.py so maybe I do not understand your question...

kaspermarstal commented:
There are some things that has not been merged yet that is needed to successfully run the HBIA experiment (on my computer at least). I still have a local copy, so will add them when I get the time. No action required on your part.
